contracts/EntryPoint.sol

```diff
@@ -191,30 +185,35 @@ contract EntryPoint is StakeManager {
     function simulateValidation(UserOperation calldata userOp) external returns (uint256 preOpGas, uint256 prefund) {
         uint256 preGas = gasleft();
 
-        bytes32 requestId = getRequestId(userOp);
-        (prefund,,) = _validatePrepayment(0, userOp, requestId);
+        UserOpInfo[] memory opInfos = new UserOpInfo[](1);
+        _validatePrepayment(0, userOp, opInfos);
+        prefund = opInfos[0].prefund;
         preOpGas = preGas - gasleft() + userOp.preVerificationGas;
 
         require(msg.sender == address(0), "must be called off-chain with from=zero-addr");
     }
 
-    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, PaymentMode paymentMode) {
-        requiredPrefund = userOp.requiredPreFund();
-        if (userOp.hasPaymaster()) {
-            paymentMode = PaymentMode.paymasterDeposit;
-        } else {
-            paymentMode = PaymentMode.walletDeposit;
-        }
+    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, address paymaster) {
+    unchecked {
+        paymaster = userOp.paymaster;
+        //when using a Paymaster, the verificationGas is used also to as a limit for the postOp call.
+        // our security model might call postOp eventually twice
+        uint256 mul = paymaster != address(0) ? 3 : 1;
```

Comment:
I am having my doubts that implicit `postOpGasLimit := verificationGasLimit` is such a great assumption.
IIRC, wallet deployment cost is counted towards the verification gas, isn't it? 
This potentially means much bigger then expected `requiredGas`, and way too much gas available to the `postOp`

Reply:
we do need SOME limit on postOp - we need verify the paymaster has "max possible gas" to pay for it.
The options are:
- constant value
- add new `UserOp.postOpGasLimit` parameter (which is used only if there's a paymaster)
- verificationGas
- callGas
- a factor (e.g. half) of one of the above.

using verificationGas is not ideal, but better than the alternatives. The only downside is that the paymaster needs a higher balance (which is never used)